File: js/popnei/bench/time_kosman_dists.mjs
/**
 * How long the Kosman distances of every pair take in WebAssembly under
 * node, and how long reading the same file alone takes.
 *
 * It is what task 3.1 of docs/plans/dists-kosman.md timed the wasm build
 * with, against the 1.43 s of "Speed" of docs/specs/dists.md, which is of
 * the calculation with the reading of the file taken out.
 * `docs/reports/dists-kosman-measurement.md` has the numbers and the load
 * averages they were taken at, and
 * `crates/popnei/benches/time_kosman_dists.py` is the same clock on the
 * native build.
 *
 * It times two passes over one source. The calculation is one call of
 * `calcPairwiseKosmanDists`, which reads every block and gives the distance
 * of every pair. The reading alone is `iterBlocks({fields: []})`, the same
 * pass with the genotypes as the only field the blocks carry, which is what
 * the calculation asks its reader for, with nothing done to the genotypes
 * but adding up how many alleles came out. The difference of the two is
 * what the calculation costs beyond the reader.
 *
 * Reading the file into a `Uint8Array` and `openVars` are outside both
 * timings, as `open_vars` is outside the native ones: `openVars` copies
 * those bytes into the memory of wasm and reads the schema and the footer,
 * and every pass reads them again from their start. What each of the two
 * took is printed under the runs. wasm has no threads, so there is one.
 *
 *     node bench/time_kosman_dists.mjs <path to a vars file> <runs>
 *
 * It is run from `js/popnei`, and it needs `npm run build` before it: that
 * is what compiles the core to WebAssembly in release and writes `wasm/`
 * and `dist/`, which the name `popnei` resolves to here. One pass of each
 * kind runs before the timed ones and is not timed, and the two kinds are
 * run one after the other inside each round, so that a machine that grows
 * busier over the runs falls on both.
 */

import { readFile } from "node:fs/promises";

import { calcPairwiseKosmanDists, init, openVars } from "popnei";

/** One call of `calcPairwiseKosmanDists`, as its pairs and its variants. */
function theCalculation(variants) {
  const distances = calcPairwiseKosmanDists(variants);
  return [distances.distVector.length, distances.passStats.numVars];
}

/**
 * One pass with the genotypes as the only field, as the alleles that came
 * out of it and the variants it took.
 *
 * The alleles are added up so that a reader which stopped filling the
 * genotypes, or filled them only when somebody read them, would show.
 */
function theReadingAlone(variants) {
  const blocks = variants.iterBlocks({ fields: [] });
  let alleles = 0;
  for (const block of blocks) {
    alleles += block.gts.length;
  }
  return [alleles, blocks.passStats.numVars];
}

/** The best, the median and the worst of `times`, in seconds. */
function saidAbout(what, times) {
  const sorted = [...times].sort((one, other) => one - other);
  const median = sorted[(sorted.length - 1) >> 1];
  return (
    `${what}: best ${sorted[0].toFixed(3)} s, ` +
    `median ${median.toFixed(3)} s, ` +
    `worst ${sorted[sorted.length - 1].toFixed(3)} s`
  );
}

const [path, runsAsText] = process.argv.slice(2);
if (path === undefined || runsAsText === undefined) {
  console.log("node bench/time_kosman_dists.mjs <path to a vars file> <runs>");
  process.exit(1);
}
const runs = Number.parseInt(runsAsText, 10);
console.log(`${path}, popnei in wasm under node, ${runs} runs, one thread`);

await init();
let started = performance.now();
const bytes = new Uint8Array(await readFile(path));
const readIn = (performance.now() - started) / 1000;
started = performance.now();
const variants = openVars(bytes);
const openedIn = (performance.now() - started) / 1000;

console.log(
  `the first run of each kind, which is not timed: ${theCalculation(variants)}`,
);
theReadingAlone(variants);

const calculations = [];
const readings = [];
for (let run = 1; run <= runs; run += 1) {
  for (const [what, passOfTheRun, times] of [
    ["the calculation", theCalculation, calculations],
    ["the reading alone", theReadingAlone, readings],
  ]) {
    started = performance.now();
    const [counted, numVars] = passOfTheRun(variants);
    const took = (performance.now() - started) / 1000;
    times.push(took);
    console.log(
      `run ${run}, ${what}: ${took.toFixed(3)} s, ${numVars} variants, ${counted}`,
    );
  }
}
console.log(saidAbout("the calculation", calculations));
console.log(saidAbout("the reading alone", readings));
const difference = Math.min(...calculations) - Math.min(...readings);
console.log(
  `the calculation with the reading taken out, on the bests: ` +
    `${difference.toFixed(3)} s`,
);
console.log(
  `reading the file into a Uint8Array, which is in neither: ` +
    `${readIn.toFixed(3)} s; openVars, which is in neither: ` +
    `${openedIn.toFixed(3)} s`,
);
variants.free();

File: crates/popnei/benches/time_kosman_dists.py
"""How long the Kosman distances of every pair take over a whole file, and
how long reading that file alone takes.

It is what task 3.1 of docs/plans/dists-kosman.md timed popnei with, against
the three numbers of "Speed" of docs/specs/dists.md, which are of the
calculation with the reading of the file taken out.
`docs/reports/dists-kosman-measurement.md` has the numbers and the load
averages they were taken at. `time_kosman_pynei.py`, beside this file, is
the same clock on pyNei, and `js/popnei/bench/time_kosman_dists.mjs` the
one on the wasm build under node.

It times two passes over the same source. The **calculation** is one call of
`calc_pairwise_kosman_dists`, which reads every block of the source and
gives the distance of every pair. The **reading alone** is
`iter_blocks(fields=())`, the same pass with the genotypes as the only field
the blocks carry, which is what the calculation asks its reader for, with
nothing done to the genotypes but adding up how many alleles came out. The
difference of the two is what the calculation costs beyond the reader, and
it is the number the targets of the spec are of.

`open_vcf` and `open_vars` are outside both timings. They read the header of
the VCF, or the schema and the footer of the vars file, and no genotype:
every pass reads the source again from its start, so the handle is built
once and given to every run. What the open itself took is printed under the
runs.

The threads are rayon's, which the library never sets: the pool takes one
thread for each core unless `RAYON_NUM_THREADS` is in the environment of the
process before the first call, which is why it is set in the command and not
from Python. Over a vars file the reader runs on the thread that calls it,
so there the threads are the calculation's alone.

    RAYON_NUM_THREADS=1 uv run python time_kosman_dists.py <path> <runs>
    uv run python time_kosman_dists.py <path> <runs>

A path that ends in `.vars` is opened with `open_vars` and anything else
with `open_vcf`. One pass of each kind runs before the timed ones and is not
timed, for the page cache. The two kinds are run one after the other inside
each round, so that a machine that grows busier over the runs falls on both.

popnei has to be built in release, `uv run maturin develop --release`: the
debug build is several times slower and says nothing about the targets.
"""

import os
import statistics
import sys
import time

from popnei import calc_pairwise_kosman_dists, open_vars, open_vcf


def the_calculation(variants) -> tuple[int, int]:
    """One call of `calc_pairwise_kosman_dists` over `variants`, as the
    distances of its pairs and the variants the pass took."""
    distances = calc_pairwise_kosman_dists(variants)
    return distances.dist_vector.shape[0], distances.pass_stats.num_vars


def the_reading_alone(variants) -> tuple[int, int]:
    """One pass over `variants` with the genotypes as the only field, as the
    alleles that came out of it and the variants the pass took.

    The alleles are added up so that a reader which stopped filling the
    genotypes, or filled them only when somebody read them, would show:
    without that count the pass would be of a column nobody touched.
    """
    blocks = variants.iter_blocks(fields=())
    alleles = sum(int(block.gts.size) for block in blocks)
    return alleles, blocks.pass_stats.num_vars


def said_about(what: str, times: list[float]) -> str:
    """The best, the median and the worst of `times`, under the name of the
    pass they are of."""
    return (
        f"{what}: best {min(times):.3f} s, median {statistics.median(times):.3f} s, "
        f"worst {max(times):.3f} s"
    )


def main() -> int:
    if len(sys.argv) != 3:
        print(__doc__)
        return 1
    path = sys.argv[1]
    runs = int(sys.argv[2])
    threads = os.environ.get("RAYON_NUM_THREADS", "one for each core")
    print(f"{path}, popnei, {runs} runs, RAYON_NUM_THREADS={threads}")

    started = time.perf_counter()
    variants = open_vars(path) if path.endswith(".vars") else open_vcf(path)
    opened_in = time.perf_counter() - started

    print(
        f"the first run of each kind, which is not timed: {the_calculation(variants)}"
    )
    the_reading_alone(variants)

    calculations: list[float] = []
    readings: list[float] = []
    for run in range(1, runs + 1):
        for what, pass_of_the_run, times in (
            ("the calculation", the_calculation, calculations),
            ("the reading alone", the_reading_alone, readings),
        ):
            started = time.perf_counter()
            counted, num_vars = pass_of_the_run(variants)
            took = time.perf_counter() - started
            times.append(took)
            print(f"run {run}, {what}: {took:.3f} s, {num_vars} variants, {counted}")
    print(said_about("the calculation", calculations))
    print(said_about("the reading alone", readings))
    print(
        f"the calculation with the reading taken out, on the bests: "
        f"{min(calculations) - min(readings):.3f} s"
    )
    print(f"opening the source, which is in neither: {opened_in:.4f} s")
    return 0


if __name__ == "__main__":
    sys.exit(main())
